File: ts/105.construct-binary-tree-from-preorder-and-inorder-traversal.ts
/*
 * @lc app=leetcode id=105 lang=typescript
 *
 * [105] Construct Binary Tree from Preorder and Inorder Traversal
 */

import TreeNode from "./types/tree-node";
// @lc code=start
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function buildTree(preorder: number[], inorder: number[]): TreeNode | null {
  let pIdx = 0;
  const iIdxMap = new Map<number, number>();

  inorder.forEach((num, i) => iIdxMap.set(num, i));

  const arrayToTree = (left: number, right: number) => {
    if (left > right) return null;

    const rootVal = preorder[pIdx];
    const root = new TreeNode(rootVal);

    pIdx++;

    root.left = arrayToTree(left, iIdxMap.get(rootVal) - 1);
    root.right = arrayToTree(iIdxMap.get(rootVal) + 1, right);

    return root;
  };

  return arrayToTree(0, preorder.length - 1);
}

// @lc code=end

File: go/105.construct-binary-tree-from-preorder-and-inorder-traversal/105.construct-binary-tree-from-preorder-and-inorder-traversal.go
/*
 * @lc app=leetcode id=105 lang=golang
 *
 * [105] Construct Binary Tree from Preorder and Inorder Traversal
 */

package leetcode

import "github.com/maguroid/leetcode/go/structures"

type TreeNode = structures.TreeNode

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func buildTree(preorder []int, inorder []int) *TreeNode {

	if len(preorder) == 0 {
		return nil
	}

	// first of preorder is absolutely root
	rootVal := preorder[0]
	// find index of root in inorder
	piv := indexOf(inorder, rootVal)

	root := &TreeNode{
		Val: rootVal,
		// leftSubtree's end is equal to inorder's root
		Left: buildTree(preorder[1:piv+1], inorder[:piv]),
		// rightSubtree is next of inorder's root
		Right: buildTree(preorder[piv+1:], inorder[piv+1:]),
	}

	return root
}

func indexOf(s []int, v int) int {
	for i, e := range s {
		if e == v {
			return i
		}
	}

	return -1
}

// @lc code=end
